### DISCUSSION TITLE
Improving testing for packaging

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2633] [General]
I had a discussion with one of Arch packagers last night who had had some issues with getting the tests to complete when they were trying to build the package.

For example, `test_imapwidget` was failing in their build environment, despite working fine on CI. 

It would be nice if there was a way that we could help the packagers and identify potential test issues before we release a new version. Obviously, we can't support every distro but one of the concerns raised was that they weren't clear what dependencies were required for the test suite.

Ideally, an Arch environment on CI would be used but I'm not sure that's supported. Alternatively, I can run builds on a machine I have at home.

Interestingly, I ran the build in a container at home last night and, for v0.18, the tests were pretty quick. However, using the latest version with the split backend testing was really slow (plus there were loads of failed tests, but that could just be a dependency issue - EDIT: yes, needed `gtk3` installed, now just the one test failure).

Interested in thoughts as to how/if we could better support packagers.

**Comment by wizardwatch** [comment]
I am going to try and grind out finishing making most of the v18 tests pass tonight for NixOS. The point of nix is reproducible builds which can help when identifying dependency/ environment issues. If you all are fine with it I can send in a pr which not only makes builds and the dev env (almost) fully reproducible but also creates a mechanism for installing the dev env for qtile on all distros nix is supported on.

**Comment by m-col** [comment]
Relevant issue: https://github.com/qtile/qtile/issues/2466. I'll take a stab at moving some of the docs around. Currently things are somewhat implied i.e. explaining the deps for X, same for Wayland, then tests are implying "on top of that...". Might be good to have an explicit table listing everything, so I'll take a look at that.

> Ideally, an Arch environment on CI would be used but I'm not sure that's supported.

Agreed - especially so we don't have to manually build a bunch of stuff due to outdated ubuntu repos. Unfortunately github actions does not support Arch. A workaround might be to use docker in ubuntu - how sure how much work that would cost/save though.

**Reply by elParaguayo** [reply]
The problem with testing on Arch is that it's a rolling release. They do provide regular builds of virtual machines so there could be some automation of retrieving the latest image and building qtile inside that. However, I agree, that's starting to feel overly complicated for the small benefit it would provide.

**Comment by elParaguayo** [comment]
So, for Arch Linux, I've taken their PKGBUILD file for v0.18 and made some changes. This is what's currently in as dependencies:

Core dependencies:
```
  'gdk-pixbuf2'
  'glibc'
  'libnotify'  # Don't think this is core, but it provides `notify-send` which we use in a test (optionally)
  'librsvg'
  'pango'
  'python-cairocffi'
  'python-cffi'
  'python-xcffib'
  'python-gobject'  # Don't think this is core either now as we removed the gi MainLoop, still needed for tests though
```

Make dependencies:
```
  'git'
  'jupyter_console'  # Is this optional? I think it's just for iqshell
  'libpulse'
  'python-ipykernel'  # As above
  'python-setuptools'
  'python-setuptools-scm'
```

Check dependencies:
```
  'graphviz'
  'gtk3'  # This is needed for the `test_window` calls
  'imagemagick'
  'libgnome-keyring'  # Not needed
  'lm_sensors'  # Not sure we actually use this in tests but it's needed for a widget
  'procps-ng'  # Provides some core apps but not sure if they're actually needed - Edit: only needed for one test (init_config for KeyboardKbdd - I should fix that...)
  'python-bowler'
  'python-dbus-next'
  'python-keyring'  # Not needed
  'python-iwlib'  # Can't imagine this is needed
  'python-psutil'  # This was mocked for tests so shouldn't be needed
  'python-pytest'
  'python-pywlroots'
  'python-xdg'
  'python-xvfbwrapper'
  'wlroots'  # Pulled in by python-pywlroots
  'xorg-xcalc'
  'xorg-xclock'
  'xorg-xeyes'
  'xorg-server-xephyr'
  'xorg-xrandr'
  'xterm'
```

I'll try an update by pulling out the things I've noted above but let me know if there's anything else that looks odd to you.

**Reply by m-col** [reply]
To avoid doing things twice, should we do another minor release as part of this whole process? We have some bug fixes and docs changes, but importantly the tests have had some changes that change the dependencies. In particular `gtk3` listed above is only needed on master and not 0.18.0, whereas xcalc, xclock, xeyes and xterm are needed for 0.18.0 but not master.

**Reply by m-col** [reply]
Also, technically the X related stuff now is optional. I have no idea how packagers might handle that change though...

**Reply by elParaguayo** [reply]
> To avoid doing things twice, should we do another minor release as part of this whole process? We have some bug fixes and docs changes, but importantly the tests have had some changes that change the dependencies. In particular `gtk3` listed above is only needed on master and not 0.18.0, whereas xcalc, xclock, xeyes and xterm are needed for 0.18.0 but not master.

I'm thinking about this for future releases i.e. can I create a PKGBUILD that successfully builds and tests master?

**Reply by elParaguayo** [reply]
> Also, technically the X related stuff now is optional. I have no idea how packagers might handle that change though...

On Arch you could have a post install script that checks to see if one of the backend's requirements are met and, if not, warn the user.

**Comment by elParaguayo** [comment]
@m-col, I'm getting some mypy related errors with the tests in the Arch build:
```
___________________________ test_check_good_key_arg ____________________________

    def test_check_good_key_arg():
>       assert check_literal_config("""
            from libqtile.config import Key
            from libqtile.command import lazy

            keys = [Key(["mod4"], "x", lazy.kill())]
            cursor_warp = True
        """)
E       assert False
E        +  where False = check_literal_config('\n        from libqtile.config import Key\n        from libqtile.command import lazy\n\n        keys = [Key(["mod4"], "x", lazy.kill())]\n        cursor_warp = Tr
ue\n    ')

test/test_check.py:86: AssertionError
----------------------------- Captured stdout call -----------------------------
error: not checking stubs due to mypy build errors:
tmpef011ai7.pyi:4: error: Cannot find implementation or library stub for module named "libqtile.config"
tmpef011ai7.pyi:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
tmpef011ai7.pyi:5: error: Cannot find implementation or library stub for module named "libqtile.layout.base"
checking qtile config file /tmp/tmpnt4h8hh9/tmpef011ai7.py
____________________________ test_extra_vars_are_ok ____________________________

    def test_extra_vars_are_ok():
>       assert check_literal_config("""
            this_is_an_extra_config_var = "yes it is"
        """)
E       assert False
E        +  where False = check_literal_config('\n        this_is_an_extra_config_var = "yes it is"\n    ')

test/test_check.py:103: AssertionError
----------------------------- Captured stdout call -----------------------------
error: not checking stubs due to mypy build errors:
tmpkg_yncry.pyi:4: error: Cannot find implementation or library stub for module named "libqtile.config"
tmpkg_yncry.pyi:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
tmpkg_yncry.pyi:5: error: Cannot find implementation or library stub for module named "libqtile.layout.base"
checking qtile config file /tmp/tmpa122ep1w/tmpkg_yncry.py
```

Any idea what's missing to make this work?

**Reply by m-col** [reply]
Huh, weird, no idea. Surely if a module is importable then mypy should be able to find its implementation?? Maybe try adding libqtile to $MYPYPATH?

**Reply by m-col** [reply]
Does the python version matter?

**Reply by elParaguayo** [reply]
I'm confused by `test_check`. I can't get it to pass locally when running in a virtual environment either.

**Reply by elParaguayo** [reply]
Do the `ignore-missing-imports = True` lines in `setup.cfg` have any relevance here? Are the tests reading the rules in that file?

**Reply by elParaguayo** [reply]
If I set `MYPYPATH` to be the root folder (i.e. the folder containing `libqtile`, `test` etc) I get different errors:
```
error: not checking stubs due to mypy build errors:
/build/qtile/src/qtile/libqtile/_ffi_pango.py:2: error: Cannot find implementation or library stub for module named "_cffi_backend"
/build/qtile/src/qtile/libqtile/_ffi_pango.py:3: error: Skipping analyzing "cairocffi._generated.ffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/_ffi_xcursors.py:2: error: Cannot find implementation or library stub for module named "_cffi_backend"
/build/qtile/src/qtile/libqtile/backend/base.py:9: error: Skipping analyzing "cairocffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/base.py:9: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
/build/qtile/src/qtile/libqtile/backend/base.py:439: error: "Static" has no attribute "x"
/build/qtile/src/qtile/libqtile/backend/base.py:440: error: "Static" has no attribute "y"
/build/qtile/src/qtile/libqtile/backend/base.py:441: error: "Static" has no attribute "width"
/build/qtile/src/qtile/libqtile/backend/base.py:442: error: "Static" has no attribute "height"
/build/qtile/src/qtile/libqtile/backend/x11/drawer.py:5: error: Skipping analyzing "cairocffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/xcbq.py:42: error: Skipping analyzing "cairocffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/xcbq.py:43: error: Skipping analyzing "cairocffi.pixbuf": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/xcbq.py:44: error: Skipping analyzing "cairocffi.xcb": found module but no type hints or library stubs
checking qtile config file /tmp/tmpt6gc8yx0/tmp9ywpftxx.py
```

**Reply by elParaguayo** [reply]
@mcol - you're the mypy guru here so hoping you have some ideas. 

`test_check.py` runs on CI but, running in an arch chroot, I get failures in four tests:
- test_check_default_config
- test_check_good_key_arg
- test_extra_vars_are_ok
- test_extra_files_are_ok

The errors are:
```
error: not checking stubs due to mypy build errors:
config.pyi:4: error: Cannot find implementation or library stub for module named "libqtile.config"
config.pyi:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
config.pyi:5: error: Cannot find implementation or library stub for module named "libqtile.layout.base"
checking qtile config file /tmp/tmpnx3ktogp/config.py
```

If I set `MYPYPATH` I get the sames tests failing but with the following messages:
```
----------------------------- Captured stdout call -----------------------------
error: not checking stubs due to mypy build errors:
/build/qtile/src/qtile/libqtile/_ffi_pango.py:2: error: Cannot find implementation or library stub for module named "_cffi_backend"
/build/qtile/src/qtile/libqtile/_ffi_pango.py:3: error: Skipping analyzing "cairocffi._generated.ffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/_ffi_xcursors.py:2: error: Cannot find implementation or library stub for module named "_cffi_backend"
/build/qtile/src/qtile/libqtile/backend/base.py:9: error: Skipping analyzing "cairocffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/base.py:9: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
/build/qtile/src/qtile/libqtile/backend/x11/drawer.py:5: error: Skipping analyzing "cairocffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/xcbq.py:42: error: Skipping analyzing "cairocffi": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/xcbq.py:43: error: Skipping analyzing "cairocffi.pixbuf": found module but no type hints or library stubs
/build/qtile/src/qtile/libqtile/backend/x11/xcbq.py:44: error: Skipping analyzing "cairocffi.xcb": found module but no type hints or library stubs
```

**Reply by elParaguayo** [reply]
A bit of progress (although I'm not sure what I've done is right...).

I added the `stubs` folder to MYPYPATH and had to make some additional stubs. I've only got two tailing tests now, both with the same error:
```
__________________________ test_check_default_config ___________________________

    def test_check_default_config():
        # the default config should always be ok
        default_config = os.path.join(os.path.dirname(__file__), '..', 'libqtile', 'resources', 'default_config.py')
>       assert run_qtile_check(default_config)
E       AssertionError: assert False
E        +  where False = run_qtile_check('/build/qtile/src/qtile/test/../libqtile/resources/default_config.py')

test/test_check.py:64: AssertionError
----------------------------- Captured stdout call -----------------------------
error: default_config failed to import: No module named 'libqtile'
Stub: at line 1
MypyFile:1(
  /tmp/tmpfh38ags_/default_config.pyi)
Runtime:
MISSING
```

Not sure how `libqtile` can be missing here!

**Comment by carlwgeorge** [comment]
Qtile was previously packaged in Fedora, but the maintainer orphaned the package and it was eventually retired.  Someone is trying to [revive it](https://bugzilla.redhat.com/show_bug.cgi?id=2025124), and I'm helping with the review.  One of the things we noticed was that installing the ["test" extra](https://github.com/qtile/qtile/blob/master/setup.cfg#L64-L67) pulls in linting dependencies, but leaves out several dependencies that are [specified in the tox config](https://github.com/qtile/qtile/blob/master/tox.ini#L26-L34).  I've sent #4084 to create separate "lint" and "test" extras with the corresponding dependencies.
